File: src/main/kotlin/ai/astha/scantheplanet/idea/scanner/model/CodeModel.kt
package ai.astha.scantheplanet.idea.scanner.model

import ai.astha.scantheplanet.idea.scanner.prompt.PromptPayload

interface CodeModel {
    val name: String
    fun analyzeChunk(prompt: PromptPayload): List<ModelFinding>
    fun analyzeChunkBatch(prompt: ai.astha.scantheplanet.idea.scanner.prompt.BatchPromptPayload): List<ModelFinding>
}

data class ModelFinding(
    val techniqueId: String? = null,
    val chunkId: String,
    val file: String,
    val startLine: Int,
    val endLine: Int,
    val severity: String,
    val confidence: Double?,
    val observation: String,
    val evidence: String,
    val reasoning: String?,
    val unknownMitigations: List<String>,
    val modelName: String
)

class CodeModelException(message: String, cause: Throwable? = null) : RuntimeException(message, cause)

File: packages/core/src/providers/CodeModel.ts
import type { ModelFinding } from '../types/Finding.js';
import type { PromptPayload, BatchPromptPayload } from '../prompt/PromptModels.js';

/**
 * Interface for code analysis models (LLM providers)
 */
export interface CodeModel {
  /** Model name/identifier */
  readonly name: string;

  /**
   * Analyze a code chunk for a single technique
   */
  analyzeChunk(prompt: PromptPayload): Promise<ModelFinding[]>;

  /**
   * Analyze a code chunk for multiple techniques (batch mode)
   */
  analyzeChunkBatch(prompt: BatchPromptPayload): Promise<ModelFinding[]>;
}

/**
 * Exception thrown by code models
 */
export class CodeModelException extends Error {
  constructor(
    message: string,
    public readonly cause?: Error
  ) {
    super(message);
    this.name = 'CodeModelException';
  }
}
